enhancements/cluster-logging/cluster-logging-es-pipeline-processing.md

```diff
@@ -0,0 +1,91 @@
+---
+title: cluster-logging-es-pipeline-processing
+
+authors:
+  - "@lvlcek"
+
+reviewers:
+  - "@alanconway"
+  - "@jcantrill"
+  - "@ewolinetz"
+
+approvers:
+  - "@alanconway"
+  - "@jcantrill"
+  - "@ewolinetz"
+
+creation-date: 2020-08-20
+last-updated: 2020-08-20
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Cluster Logging Elasticsearch Pipeline Processing Of Indexed Documents
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal opens possibility to route individual documents (logs) into different Elasticsearch indices
+based on data driven criteria.  Under the hood it utilizes Elasticsearch Ingesting [Pipeline Definition](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/pipeline.html) to dynamically inspect incoming documents and apply appropriate processor(s) such as [parsing JSON data](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/json-processor.html) in document field or pointing
```

Comment:
do we know what sort of overhead this adds to ES? will we want to scale a specific type of node up higher (like client) to help deal with the additional workloads?

Reply:
Could we also have a pipeline to create our rollover indices and aliases for us?

Reply:
> I would hesitate to start writing our own ES plugin again based on our previous experience

This is valid concern but I do not think our previous experience with maintaining plugin that deal with ES security layers fully translates here. Our experience with [Prometheus plugin](https://github.com/vvanholl/elasticsearch-prometheus-exporter) has been a lot more smooth. The only real pain that is left is "packaging" of plugins but that is _our own problem_.

There is valid use for 3rd party plugins as ingest processors and [this is officially supported](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/ingest-processors.html#_ingest_processor_plugins). In this case all we need to do is take existing Apache 2 licensed ES code of [DateIndexNameProcessor.java](https://github.com/elastic/elasticsearch/blob/v6.8.12/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateIndexNameProcessor.java) (the [date index name processor](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/date-index-name-processor.html)) and update it slightly to enable constructing name of index that will fit our index naming schema. IMO nothing really complicated...

Reply:
> Aliasing and index management can be defined now using the tools that exist in the EO. Can we make use of not only our existing apis or es apis until we get to a later version of OD? Does it maybe support our requirements? I would hesitate to start writing our own ES plugin again based on our previous experience

I wasn't looking to replace it in the near term future, it was more of a general question whether or not it could be possible in the future to do.

Reply:
> Could we also have a pipeline to create our rollover indices and aliases for us?

This is a bit orthogonal to this proposal. We can think about other applications of pipeline processing but let's do it later once we have more experience with it. I would rather move to ES 7.x which will give us overall better index management.